// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

syntax = "proto3";

package spoor;

import "google/protobuf/timestamp.proto";

// Note: Data types should match the definitions in
// //spoor/runtime/trace/trace.h.

message FunctionInfo {
  optional string linkage_name = 1;
  optional string demangled_name = 2;
  optional string file_name = 3;
  optional string directory = 4;
  optional int32 line = 5;
  optional bool instrumented = 6;
};

message InstrumentedFunctionMap {
  optional string module_id = 1;
  map<uint64, FunctionInfo> function_map = 2;
  optional google.protobuf.Timestamp created_at = 3;
}

message Event {
  enum Type {
    FUNCTION_EXIT = 0;
    FUNCTION_ENTRY = 1;
  }

  uint64 session_id = 1;
  int64 process_id = 2;
  uint64 thread_id = 3;
  uint64 function_id = 4;
  Type type = 5;
  google.protobuf.Timestamp timestamp = 6;
}

message Trace {
  repeated Event events = 1;
  repeated string modules = 2;
  map<uint64, FunctionInfo> function_map = 3;
  google.protobuf.Timestamp created_at = 4;
}
